qa/lucene-index-compatibility/src/javaRestTest/java/org/elasticsearch/lucene/FullClusterRestartLuceneIndexCompatibilityIT.java

```diff
@@ -374,4 +387,118 @@ public void testRestoreIndexOverClosedIndex() throws Exception {
+            assertThat(indexVersion(asyncSearchIndex, true), equalTo(VERSION_MINUS_2));
+            return;
+        }
+
+        if (isFullyUpgradedTo(VERSION_MINUS_1)) {
+            // check .async-search index is readable
+            assertThat(indexVersion(asyncSearchIndex, true), equalTo(VERSION_MINUS_2));
+
+            // migrate system indices
+            Request migrateRequest = new Request("POST", "/_migration/system_features");
+            assertThat(
+                ObjectPath.createFromResponse(client().performRequest(migrateRequest)).evaluate("features.0.feature_name"),
+                equalTo("async_search")
+            );
+            assertBusy(() -> {
+                Request checkMigrateProgress = new Request("GET", "/_migration/system_features");
+                Response resp = null;
+                try {
+                    assertFalse(
+                        ObjectPath.createFromResponse(client().performRequest(checkMigrateProgress))
+                            .evaluate("migration_status")
+                            .equals("IN_PROGRESS")
+                    );
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+
+            // check search results from n-2 search are still readable
+            checkRetrieveAsyncSearch(async_search_ids.get("n-2_id"), numDocs);
+
+            // perform new async search and check its readable
+            String asyncId = searchAsyncAndStoreId(asyncSearchRequest, "n-1_id");
+            checkRetrieveAsyncSearch(asyncId, numDocs);
+            assertBusy(() -> assertDocCountNoWarnings(client(), asyncSearchIndex, 2));
+
+            // in order to move to current version we need write block for n-2 index
+            addIndexBlock(index, IndexMetadata.APIBlock.WRITE);
+        }
+
+        if (isFullyUpgradedTo(VERSION_CURRENT)) {
+            assertThat(indexVersion(index, true), equalTo(VERSION_MINUS_2));
+            checkRetrieveAsyncSearch(async_search_ids.get("n-2_id"), numDocs);
+            checkRetrieveAsyncSearch(async_search_ids.get("n-1_id"), numDocs);
+
+            // check system index is still writeable
+            String asyncId = searchAsyncAndStoreId(asyncSearchRequest, "n_id");
+            checkRetrieveAsyncSearch(asyncId, numDocs);
+            assertBusy(() -> assertDocCountNoWarnings(client(), asyncSearchIndex, 3));
+        }
+
+    }
+
+    private static String searchAsyncAndStoreId(Request asyncSearchRequest, String asyncIdName) throws IOException {
+        ObjectPath resp = ObjectPath.createFromResponse(client().performRequest(asyncSearchRequest));
+        String asyncId = resp.evaluate("id");
+        assertNotNull(asyncId);
+        async_search_ids.put(asyncIdName, asyncId);
+        return asyncId;
+    }
+
+    private static void checkRetrieveAsyncSearch(String asyncId, int numDocs) throws IOException {
```

Comment:
Maybe rename to something like:
```suggestion
    private static void assertAsyncSearchHitCount(String asyncId, int numDocs) throws IOException {
```